Es6 plugin: Search

iisa opened the pull request:
#### Test Steps
- pull branch
- fresh module install, cmd: `npm install`
- run server: `npm run serve`
- Test desktop
  - go to: http://127.0.0.1:8000/BookReaderDemo/demo-ia-olivertwist.html
  - search text using input: `child` (wait a while, then wait a lil more) -->  results markers successfully pop up
  - hover over results marker. --> text tooltip pops up
    - known: page highlight bleeds over tooltip
  - hover off of results marker --> tooltip disappears
  - click on marker --> takes you to page with results highlighted
  - refresh page that is full of search results (wait for it...) --> results markers show
  - search a new word: `shmeowza` --> no results window pops up & fades out
- Run tests on mobile device (browser dev emulator)
  - go to: http://127.0.0.1:8000/BookReaderDemo/demo-ia-olivertwist.html
  - open 🍔 --> search section is visible
 - search text using input: `child` (wait a while, then wait a lil more) -->  results & count appear below input
  - click on search result -> takes you to page
 - refresh page --> results are shown in the search drawer
  - search a new word: `shmeowza` --> no results are displayed


internal ticket: https://webarchive.jira.com/browse/WEBDEV-3206

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,575 @@
+        return true;
+      }
+    }
+  }
```
```suggestion
  results.matches.forEach(match => {
    match.par[0].boxes.forEach(box =>
      const pageIndex = this.leafNumToIndex(box.page);
      if (jQuery.inArray(pageIndex, visiblePages) >= 0) {
        return true;
      }
    });
  });
```

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,575 @@
+     * @type { bool } - (default=false) disablePopup - don't show the modal progress
+     */
+    disablePopup: false
+  };
```
Let's not remove the `error` and `success` options; those are still public options, so we don't know if someone who uses BR outside of ia is using them. We need to try to keep things as backwards compatible as possible, because we're not providing notice of breaking changes. This would require a v5 version bump.

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,575 @@
+/**
+ * Search Results return handler
+ * @param { object } results
+ * @param { array } results.matches
```
Oh, last thing; let's give these some `typedef`s so that we can reference their types, and add info about what fields are present. I think that would help make things much clearer!

iisa replied:
yes I like. :excellent:

iisa commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,575 @@
+ * Main search results error handler
+ * @param { array } results
+ */
+BookReader.prototype.BRSearchCallbackError = function(results) {
```
thanks for catching! I was deep in consolidating..

nsharma123 commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,556 @@
+  }
+
+  const processSearchResults = (data) => {
+    if (data.error || 0 == data.matches.length) {
```
Per es6 standards, you can prevent `Yoda condition` where the literal value coming first and variable is second. So it can be write something like `data.matches.length === 0` (including strict type with ===). Thanks

iisa replied:
thanks, updated.

nsharma123 commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,556 @@
+/**
+ * Returns true if a search highlight is currently being displayed
+ */
+BookReader.prototype.searchHighlightVisible = function() {
```
lets add jsdocs include `@return { ... }`

iisa replied:
thanks for catching. added

iisa commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,599 @@
+ * @property {boolean} disablePopup
+ * @property {(null|function)} error - @deprecated at v.5.0
+ * @property {(null|function)} success - @deprecated at v.5.0
+ */
```
new `@typedef`: Search Options
We pass this around to success & error callbacks, so it was helpful to make one for it.
I abstracted it up to the "top level," vscode couldn't back link the references to this when I added it right above the callback inside the function

iisa commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,599 @@
+ * @property {string} error
+ * @property {array} matches
+ * @property {boolean} indexed
+ */
```
new @typedef: `searchInsideResults`
This also gets passed to the varied callbacks we have.

cdrini replied:
This should get all of them :+1:

```suggestion
/**
  * @typedef {object} SearchInsideMatchBox
  * @property {number} page
  * @property {number} r
  * @property {number} l
  * @property {number} b
  * @property {number} t
  * @property {HTMLDivElement} [div]
  */

/**
 * @typedef {object} SearchInsideMatch
 * @property {string} text
 * @property {Array<{ page: number, boxes: SearchInsideMatchBox[] }>} par
 */

/**
 * @typedef {object} SearchInsideResults
 * @property {string} error
 * @property {SearchInsideMatch[]} matches
 * @property {boolean} indexed
 */
```

iisa commented:
Hi @cdrini, I addressed the suggestions you had, thanks!
I will hold here though, because some of the changes we introduced in the for loop updates started breaking the functionality.

This is a great place to leave this file.

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,599 @@
+ * @param {object} results
+
+ * @param {jQuery} $el
+ * @param {boolean} fade
```
```suggestion
 * @param {SearchInsideResults} results
 * @param {jQuery} $el
 * @param {boolean} fade
```

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,599 @@
+/**
+ * Search Results return handler
+ * @callback
+ * @param {searchInsideResults} results
```
```suggestion
 * @param {SearchInsideResults} results
```

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,599 @@
+/**
+ * Main search results error handler
+ * @callback
+ * @param {searchInsideResults} results
```
```suggestion
 * @param {SearchInsideResults} results
```

cdrini commented on src/js/plugins/plugin.search.js:
```diff
@@ -0,0 +1,599 @@
+    } else {
+      hasCustomSuccess
+        ? options.success(searchInsideResults, options)
+        : this.BRSearchCallback(searchInsideResults, options);
```
For backwards compatibility
```suggestion
        ? options.error.call(this, searchInsideResults, options)
        : this.BRSearchCallbackError(searchInsideResults, options);
    } else {
      hasCustomSuccess
        ? options.success.call(this, searchInsideResults, options)
        : this.BRSearchCallback(searchInsideResults, options);
```

iisa replied:
thanks! that is important 📝
